javascript/ql/src/experimental/semmle/javascript/Actions.qll

```diff
@@ -0,0 +1,314 @@
+    string version;
+
+    Uses() {
+      step.lookup("uses") = this and
+      // Simple regular expression to split up an Action reference `owner/repo@version` into its components.
+      exists(string regexp | regexp = "([^/]+)/([^/@]+)@(.+)" |
+        repositoryOwner = this.getValue().regexpCapture(regexp, 1) and
+        repositoryName = this.getValue().regexpCapture(regexp, 2) and
+        version = this.getValue().regexpCapture(regexp, 3)
+      )
+    }
+
+    /** Gets the step this field belongs to. */
+    Step getStep() { result = step }
+
+    /** Gets the owner and name of the repository where the Action comes from, e.g. `actions/checkout` in `actions/checkout@v2`. */
+    string getGitHubRepository() { result = repositoryOwner + "/" + repositoryName }
+
+    /** Gets the version reference used when checking out the Action, e.g. `v2` in `actions/checkout@v2`. */
+    string getVersion() { result = version }
+  }
+
+  /**
+   * A `with` field within an Actions job step, which references an action as a reusable unit of code.
+   * See https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepswith.
+   *
+   * For example:
+   * ```
+   * with:
+   *   arg1: 1
+   *   arg2: abc
+   * ```
+   */
+  class With extends YAMLNode, YAMLMapping {
+    Step step;
+
+    With() { step.lookup("with") = this }
+
+    /** Gets the step this field belongs to. */
+    Step getStep() { result = step }
+  }
+
+  /**
+   * A `ref:` field within an Actions `with:` specific to `actions/checkout` action.
+   *
+   * For example:
+   * ```
+   * uses: actions/checkout@v2
+   * with:
+   *   ref: ${{ github.event.pull_request.head.sha }}
+   * ```
+   */
+  class Ref extends YAMLNode, YAMLString {
+    With with;
+
+    Ref() { with.lookup("ref") = this }
+
+    With getWith() { result = with }
+  }
+
+  /**
+   * A `run` field within an Actions job step, which runs command-line programs using an operating system shell.
+   * See https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsrun.
+   */
+  class Run extends YAMLNode, YAMLString {
+    Step step;
+
+    Run() { step.lookup("run") = this }
+
+    /** Gets the step that executes this `run` command. */
+    Step getStep() { result = step }
+
+    /**
+     * Holds if `${{ e }}` is a GitHub Actions expression evaluated within this `run` command.
+     * See https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions.
+     */
+    string getAReferencedExpression() {
+      // We use `regexpFind` to obtain *all* matches of `${{...}}`,
+      // not just the last (greedy match) or first (reluctant match).
+      // TODO: This only handles expression strings that refer to contexts.
+      // It does not handle operators within the expression.
+      result =
+        this.getValue().regexpFind("(?<=\\$\\{\\{\\s*)[A-Za-z0-9_\\.\\-]+(?=\\s*\\}\\})", _, _)
```

Comment:
```suggestion
        this.getValue().regexpFind("(?:\\$\\{\\{\\s*)[A-Za-z0-9_\\.\\-]+(?=\\s*\\}\\})", _, _)
```

I think you can just replace the positive lookbehind with a non-capturing group.  
Because the regexp does error when I try to input it into some Java regexp parsers: https://regex101.com/r/ClMFh9/1/  
So it might be that it fails depending on what JDK you are running :shrug:

Reply:
That would break the regex. I just removed the lookbehind completely. @adityasharad what do you think?

Reply:
That will match a string ending with `}}` even if you don't have the leading `${{`.
Might be ok. If you want to keep the behaviour of extracting only the section between the braces, together with Erik's suggestion for a regexp without lookbehind, a somewhat ugly trick is to do a second regexp capture as follows:
```ql
result = this.getValue().regexpFind("(?:\\$\\{\\{\\s*)[A-Za-z0-9_\\.\\-]+(?=\\s*\\}\\})", _, _).regexpCapture("\\$\\{\\{\\s*(.*)", 1)
```

Reply:
Thanks!